### PR TITLE
AZ-style policy head support

### PR DESCRIPTION
**Pull request by Ttl**
Policy head from the final AZ paper. Head is one 256x256 3x3 convolution with batch norm and relu followed by 256x80 3x3 convolution. 80x8x8 policy stack is then re-indexed to match the existing 1858 element policy vector.

az_pol_tensorboard2

Red is AZ-style policy head and the green is the current one.

~CCRL dataset trained 20x256 network with this policy head: http://hforsten.com/leelaz/256x20-az-pol-map-200000.pb.gz~ (Obsolote due to protobuf changes)

Requires protobuf changes from: https://github.com/LeelaChessZero/lczero-common/pull/8

### CODE REVIEW

File: src/neural/cuda/network_cudnn.cc [hunk 1]
```diff
@@ -193,7 +196,11 @@ class CudnnNetwork : public Network {
       processConvBlock(weights.residual[i].conv1, true, 3);
       processConvBlock(weights.residual[i].conv2, true, 3);
     }
-    processConvBlock(weights.policy, true, 1);
+    if (conv_policy_) {
+      processConvBlock(weights.policy1, true, 1);
+    } else {
+      processConvBlock(weights.policy, true, 1);
```
[the remark this answers is not in the document]
Comment (Ttl) [reply src/neural/cuda/network_cudnn.cc]: `processConvBlock` merges batch norm parameters to weights and biases. With convolutional policy the second convolution doesn't have batch norm so this function shouldn't be called and in old policy head the first convolution doesn't exist.

**Comment by Ttl** [comment]
Some test results with this PR. 5man TBs were used and all other parameters were defaults.

CCRL dataset trained network at 200k steps against CCRL trained net with old policy head (20x256 SE, 80 policy head channels):
```
   # PLAYER              :  RATING  ERROR  POINTS  PLAYED   (%)  CFS(%)
   1 lc0_az_pol_200k     :    29.9    7.9   434.0     800    54     100
   2 lc0_80ch_200k       :     0.0    7.9   366.0     800    46     ---

TC: 10s+0.5s, GTX 1080 Ti.
```

File: src/neural/cuda/network_cudnn.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tilps) [reply src/neural/cuda/network_cudnn.cc]: Aha - maybe a comment to make that clearer?

**Comment by Ttl** [comment]
One more test with 80k step trained nets (end of the first LR) at 800 nodes per move:

```
   # PLAYER            :  RATING  ERROR  POINTS  PLAYED   (%)  CFS(%)
   1 lc0_az_pol_80k    :    88.6    8.1   499.0     800    62     100
   2 lc0_ccrl_80k      :     0.0    8.1   301.0     800    38     ---
```

**Comment by Ttl** [comment]
Needs modifications for protobuf changes in: https://github.com/LeelaChessZero/lczero-common/pull/8

**Comment by Ttl** [comment]
Protobuf changes done. Here is one test network: http://hforsten.com/leelaz/256x20-az-pol-map-200010.pb.gz

File: src/neural/cuda/layers.cc [hunk 2]
```diff
@@ -564,6 +564,60 @@ FCLayer<DataType>::~FCLayer() {
...
+void PolicyMapLayer<DataType>::LoadWeights(const short* cpuWeight,
+                                           void* /*scratch*/) {
+  size_t weight_size = sizeof(short) * usedSize;
+
+  if (std::is_same<half, DataType>::value) {
+    // convert CHW to HWC
+    int C = usedSize / 64;
+    int Cin = this->input_->GetC();
+    usedSize = Cin * 64;
+    short* convertedWeights = new short[usedSize];
```
[the remark this answers is not in the document]
Comment (Ttl) [reply src/neural/cuda/layers.cc]: cudnn-fp16 part was done by @ankan-ban. I don't have a GPU to test it.

**Comment by Ttl** [comment]
Code for generating the map is part of the same code that generates it for the training code. I'll put it to the training repository.

I have verified that OpenCL, blas and cudnn outputs match. I also verified the generated mapping by replacing the mapping kernel with the same matrix multiplication that the training code uses and got an exact match for visit distribution of deep search. Tests also show Elo gain so everything should be working correctly.

File: src/neural/cuda/common_kernels.cu [hunk 3]
```diff
@@ -466,6 +466,36 @@ void globalScale(int N, int C, T* output, const T* input, const T* scaleBias,
...
+  int j = indices[i];
+
+  if (j >= 0) {
+    output[n * outputSize + j] = input[n * inputSize + i];
+  }
+}
+
+template <typename T>
+void PolicyMap(int N, T* output, const T* input, const short* indices,
+        int inputSize, int usedSize, int outputSize) {
```
Comment (Tilps) [inline src/neural/cuda/common_kernels.cu]: Formatting in this file doesn't look like its had the auto formatter run over it (possibly because of file extension).

File: src/neural/cuda/kernels.h [hunk 4]
```diff
@@ -76,6 +76,10 @@ void Se_Fp16_NHWC(int N, int C, int numFc1Out, half* output, const half* skip,
                   const half* input, const half* w1, const half* b1,
                   const half* w2, const half* b2, const half* bPrev);
 
+template <typename T>
+void PolicyMap(int N, T* output, const T* input, const short* indices,
+        int inputSize, int usedSize, int outputSize);
```
Comment (Tilps) [inline src/neural/cuda/kernels.h]: Indent should be deeper for this style of wrap.

File: src/neural/cuda/layers.cc [hunk 5]
```diff
@@ -564,6 +564,60 @@ FCLayer<DataType>::~FCLayer() {
...
+    delete[] convertedWeights;
+  } else {
+    ReportCUDAErrors(cudaMemcpyAsync(weights_, cpuWeight, weight_size,
+                                     cudaMemcpyHostToDevice));
+  }
+}
+
+template <typename DataType>
+void PolicyMapLayer<DataType>::Eval(int N, DataType* output_tensor,
+                          const DataType* input_tensor, const DataType* /*input2*/,
```
Comment (Tilps) [inline src/neural/cuda/layers.cc]: This wrapping doesn't look standard - maybe rerun autoformat?

File: src/neural/shared/policy_map.h [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+
+ You should have received a copy of the GNU General Public License
+ along with Leela Chess.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+namespace lczero {
+
+const short kConvPolicyMap[] = {\
```
Comment (mooskagh) [inline src/neural/shared/policy_map.h]: No need for backslash. :)
And `int16_t` is probably more portable for this use case than `short`.
